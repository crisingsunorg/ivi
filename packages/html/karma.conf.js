const webpack = require("webpack");

module.exports = function (config) {
  config.set({
    browsers: ["ChromeHeadless"],
    frameworks: ["mocha"],
    files: ["__tests__/index.ts"],
    preprocessors: { "__tests__/index.ts": ["webpack", "sourcemap"] },

    webpack: {
      module: {
        rules: [
          {
            test: /\.ts$/,
            exclude: /node_modules/,
            use: [
              {
                loader: "ts-loader",
                options: {
                  configFileName: "tsconfig.json",
                },
              },
            ],
          },
        ],
      },
      plugins: [
        new webpack.DefinePlugin({
          "__IVI_DEV__": true,
          "__IVI_BROWSER__": true,
        }),
        new webpack.SourceMapDevToolPlugin({
          test: /\.(ts|js)$/,
        }),
      ],
      resolve: {
        extensions: [".ts", ".js"],
      },
      performance: {
        hints: false
      },
    },

    webpackMiddleware: {
      stats: "errors-only",
      noInfo: true
    },

    reporters: ["mocha"],

    colors: true,
    autoWatch: true,

    client: {
      mocha: {
        reporter: "html",
        ui: "bdd",
      }
    },

    mime: {
      "text/x-typescript": ["ts"],
    },
  });
};
