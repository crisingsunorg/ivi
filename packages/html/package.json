{
  "name": "ivi-html",
  "main": "dist/cjs/index.js",
  "module": "dist/js/index.js",
  "typings": "dist/typings/index.d.ts",
  "version": "0.1.0",
  "description": "Javascript (TypeScript) library for building web user interfaces.",
  "license": "MIT",
  "author": {
    "name": "Boris Kaul",
    "email": "localvoid@gmail.com",
    "url": "https://github.com/localvoid"
  },
  "keywords": [
    "ui",
    "virtual",
    "dom",
    "virtualdom",
    "vdom",
    "diff",
    "browser",
    "typescript"
  ],
  "homepage": "https://github.com/ivijs/ivi",
  "repository": {
    "type": "git",
    "url": "https://github.com/ivijs/ivi"
  },
  "bugs": {
    "url": "https://github.com/ivijs/ivi/issues"
  },
  "github": "https://github.com/ivijs/ivi",
  "files": [
    "dist",
    "src",
    "README.md",
    "LICENSE"
  ],
  "scripts": {
    "prepublish": "npm run dist",
    "dist": "npm run dist:es6 && npm run dist:cjs",
    "dist:es6": "../../node_modules/.bin/tsc --project ./tsconfig.build.es6.json",
    "dist:cjs": "../../node_modules/.bin/tsc --project ./tsconfig.build.cjs.json",
    "test": "../../node_modules/.bin/karma start --single-run",
    "test:watch": "../../node_modules/.bin/karma start",
    "lint": "../../node_modules/.bin/tslint --project ./tsconfig.json"
  },
  "dependencies": {},
  "peerDependencies": {
    "ivi-core": "^0.1.0",
    "ivi": "^0.8.0"
  },
  "devDependencies": {
    "ivi-core": "^0.1.0",
    "ivi": "^0.8.0"
  }
}