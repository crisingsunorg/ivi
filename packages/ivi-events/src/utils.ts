import { FEATURES, FeatureFlags } from "ivi-core";
import { NativeEventSourceFlags } from "./flags";
import { EventHandler } from "./event_handler";

/**
 * Extends Element interface with a special property that is used to assign synthetic event handlers list.
 */
declare global {
  interface Element {
    _ev: Array<EventHandler | null> | EventHandler | null | undefined;
  }
}

/**
 * getEventTarget retrieves target element from an event.
 *
 * There are still some differences in modern browsers, so we need to use this function to retrieve event targets.
 *
 * #quirks
 *
 * @param ev Event.
 * @returns Target Element.
 */
export function getEventTarget(ev: Event): EventTarget {
  let target = ev.target || window;

  /**
   * Some browsers are implementing it according to SVG 1.1 specs:
   *
   * https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/7998724/
   *
   * "If event attributes are assigned to referenced elements, then the actual target for the event will be the
   * SVGElementInstance object within the 'instance tree' corresponding to the given referenced element."
   *
   * https://www.w3.org/TR/SVG/struct.html#UseElement
   *
   * SVG 2 redefined the use event handling model:
   *
   * - Removed the SVGElementInstance and SVGElementInstanceList interfaces, and the corresponding attributes on the
   *   SVGUseElement interface.
   * - Changed the ‘use’ element event flow to follow the Shadow DOM spec.
   *
   * https://www.w3.org/TR/SVG2/changes.html#structure
   */
  if ((target as any).correspondingUseElement !== undefined) {
    target = (target as any).correspondingUseElement;
  }

  /**
   * Safari fires events on Text Nodes.
   *
   * http://www.quirksmode.org/js/events_properties.html
   */
  return (target as Node).nodeType === 3 ? (target as Node).parentNode! : target;
}

/**
 * `{ capture: true, passive: true }` object that should be used as a 3rd parameter for `addEventListener` method.
 */
export const EVENT_CAPTURE_PASSIVE_OPTIONS =
  ((FEATURES & FeatureFlags.PassiveEvents) !== 0) ?
    { "capture": true, "passive": true } :
    true;

/**
 * `{ passive: true }` object that should be used as a 3rd parameter for `addEventListener` method.
 */
export const EVENT_PASSIVE_OPTIONS =
  ((FEATURES & FeatureFlags.PassiveEvents) !== 0) ?
    { "passive": true } :
    false;

/**
 * getNativeEventOptions converts `NativeEventSourceFlags` to event options that can be used as a 3rd parameter
 * for `addEventListener` method.
 *
 * #quirks
 *
 * @param flags See `NativeEventSourceFlags` for details.
 * @returns Option object that can be used as a 3rd parameter for `addEventListener` method.
 */
export function getNativeEventOptions(
  flags: NativeEventSourceFlags,
): boolean | { capture?: boolean, passive?: boolean } {
  if ((flags & NativeEventSourceFlags.Passive) !== 0) {
    if ((flags & NativeEventSourceFlags.Capture) !== 0) {
      return EVENT_CAPTURE_PASSIVE_OPTIONS;
    }
    return EVENT_PASSIVE_OPTIONS;
  }
  if ((flags & NativeEventSourceFlags.Capture) !== 0) {
    return true;
  }
  return false;
}

/**
 * setEventHandlersToDOMNode assigns a event handlers to the DOM node.
 *
 * @param node DOM Node.
 * @param events Event Handlers.
 */
export function setEventHandlersToDOMNode(
  node: Element,
  events: Array<EventHandler | null> | EventHandler | null,
): void {
  node._ev = events;
}

/**
 * getEventHandlersFromDOMNode retrieves event handlers from the DOM node.
 *
 * @param node DOM Node.
 * @returns Event Handlers.
 */
export function getEventHandlersFromDOMNode(
  node: Element,
): Array<EventHandler | null> | EventHandler | undefined | null {
  return node._ev;
}
