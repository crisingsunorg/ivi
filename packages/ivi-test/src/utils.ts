import { EventHandler, EventSource } from "ivi-events";
import { VNode, VNodeFlags } from "ivi";

/**
 * containsClassName checks if className list contains a className.
 *
 * @param classNames className list in a HTML format with all values separated by space characters.
 * @param className className to check.
 * @returns true when className list contains a className.
 */
export function containsClassName(classNames: string, className: string): boolean {
  return classNames
    .split(" ")
    .map((c) => c.trim())
    .some((c) => c === className);
}

/**
 * matchValues checks property values for matching. Properties that aren't in `match` property are ignored.
 *
 * @param props Properties that should be matched against `match` property.
 * @param match Is a map of values that should match.
 * @returns true when properties are matching.
 */
export function matchValues(
  props: { [key: string]: any } | null,
  match: { [key: string]: any } | null,
): boolean {
  if (match !== null) {
    const keys = Object.keys(match);
    if (keys.length > 0) {
      if (props === null) {
        return false;
      }
      for (let i = 0; i < keys.length; i++) {
        const key = keys[i];
        if (props.hasOwnProperty(key) === false || props[key] !== match[key]) {
          return false;
        }
      }
    }
  }
  return true;
}

/**
 * matchKeys checks that property keys are existing.
 *
 * @param props Properties that should be checked against `match` property.
 * @param match Is a map of keys that should exist.
 * @returns true when matching properties are existing.
 */
export function matchKeys(
  props: { [key: string]: any } | null,
  match: { [key: string]: boolean },
): boolean {
  const keys = Object.keys(match);
  if (keys.length > 0) {
    if (props === null) {
      return false;
    }
    for (let i = 0; i < keys.length; i++) {
      const key = keys[i];
      if (match[key] !== props.hasOwnProperty(key)) {
        return false;
      }
    }
  }
  return true;
}

/**
 * containsEventHandler checks if event handler list contain an event source.
 *
 * @param eventHandlers Event handlers.
 * @param eventSource Event source.
 * @returns true when event handlers contain an event source.
 */
export function containsEventHandler(
  eventHandlers: Array<EventHandler | null> | EventHandler | null,
  eventSource: EventSource,
): boolean {
  if (eventHandlers !== null) {
    if (Array.isArray(eventHandlers)) {
      for (let i = 0; i < eventHandlers.length; i++) {
        const h = eventHandlers[i];
        if (h !== null && h.source === eventSource) {
          return true;
        }
      }
    } else {
      return (eventHandlers.source === eventSource);
    }
  }
  return false;
}

const VNodeLooseMatchFlags = 0
  | VNodeFlags.Text
  | VNodeFlags.Element
  | VNodeFlags.ComponentFunction
  | VNodeFlags.ComponentClass
  | VNodeFlags.InputElement
  | VNodeFlags.TextAreaElement
  | VNodeFlags.MediaElement
  | VNodeFlags.SvgElement
  | VNodeFlags.Connect
  | VNodeFlags.UpdateContext
  | VNodeFlags.KeepAlive
  | VNodeFlags.VoidElement;

/**
 * isVNodeLooseMatch performs a loose match on VNodes.
 *
 * @param a VNode.
 * @param b VNode.
 * @returns true when VNode are loosely matching.
 */
export function isVNodeLooseMatch(a: VNode<any>, b: VNode<any>): boolean {
  const bFlags = b._flags;
  if (((a._flags ^ bFlags) & VNodeLooseMatchFlags) !== 0) {
    return false;
  }

  if (a._tag !== b._tag) {
    return false;
  }

  if ((bFlags & VNodeFlags.Element) !== 0) {
    if (b._props !== null) {
      if (a._props === null) {
        return false;
      }

      if (matchValues(a._props.attrs, b._props.attrs) === false) {
        return false;
      }
      if (matchValues(a._props.syle, b._props.style) === false) {
        return false;
      }

    }
  }
  return true;
}
