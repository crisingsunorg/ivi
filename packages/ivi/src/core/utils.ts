/**
 * mut mutates an object.
 *
 * @example
 *
 *     interface Entry {
 *       _dirty: number;
 *       title: string;
 *     }
 *     const m = (entry: Entry) => { entry._dirty = dirty(); };
 *     const useEntry = selector((entry: Entry) => entry._dirty);
 *     const entrySetTitle = mut(m, (entry: Entry, title: string) => { entry.title = title; });
 *
 *     const EntryView = component<Entry>((c) => {
 *       const dirtyCheckEntry = useEntry(c);
 *
 *       return (entry) => (
 *         dirtyCheckEntry(entry),
 *
 *         div().t(entry.title)
 *       )
 *     });
 *
 * @param s - Selector function.
 * @param shouldUpdate - Should update function.
 * @returns selector factory.
 */
export const mut = <T, U extends any[], V>(d: (v: T, flags?: V) => void, fn: (v: T, ...args: U) => void | V) => (
  function () {
    const r = fn.apply(void 0, arguments);
    if (r === void 0 || r > 0) {
      d(arguments[0], r);
    }
  } as (v: T, ...args: U) => void | V
);
