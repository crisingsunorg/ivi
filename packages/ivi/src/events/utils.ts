import { PASSIVE_EVENTS } from "ivi-core";

/* istanbul ignore next */
/**
 * getEventTarget retrieves target element from an event.
 *
 * There are still some differences in modern browsers, so we need to use this function to retrieve event targets.
 *
 * #quirks
 *
 * @param ev Event.
 * @returns Target Element.
 */
export function getEventTarget(ev: Event): EventTarget {
  let target = ev.target || window;

  /**
   * Some browsers are implementing it according to SVG 1.1 specs:
   *
   * https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/7998724/
   *
   * "If event attributes are assigned to referenced elements, then the actual target for the event will be the
   * SVGElementInstance object within the 'instance tree' corresponding to the given referenced element."
   *
   * https://www.w3.org/TR/SVG/struct.html#UseElement
   *
   * SVG 2 redefined the use event handling model:
   *
   * - Removed the SVGElementInstance and SVGElementInstanceList interfaces, and the corresponding attributes on the
   *   SVGUseElement interface.
   * - Changed the ‘use’ element event flow to follow the Shadow DOM spec.
   *
   * https://www.w3.org/TR/SVG2/changes.html#structure
   */
  if ((target as any).correspondingUseElement !== undefined) {
    target = (target as any).correspondingUseElement;
  }

  return target;
}

/* istanbul ignore next */
/**
 * `{ capture: true, passive: true }` object that should be used as a 3rd parameter for `addEventListener` method.
 */
export const EVENT_CAPTURE_PASSIVE_OPTIONS = PASSIVE_EVENTS ? { "capture": true, "passive": true } : true;

/* istanbul ignore next */
/**
 * `{ capture: true, passive: false }` object that should be used as a 3rd parameter for `addEventListener` method.
 */
export const EVENT_CAPTURE_ACTIVE_OPTIONS = PASSIVE_EVENTS ? { "capture": true, "passive": false } : true;

/* istanbul ignore next */
/**
 * `{ passive: true }` object that should be used as a 3rd parameter for `addEventListener` method.
 */
export const EVENT_PASSIVE_OPTIONS = PASSIVE_EVENTS ? { "passive": true } : false;

/* istanbul ignore next */
/**
 * `{ passive: false }` object that should be used as a 3rd parameter for `addEventListener` method.
 */
export const EVENT_ACTIVE_OPTIONS = PASSIVE_EVENTS ? { "passive": false } : false;
