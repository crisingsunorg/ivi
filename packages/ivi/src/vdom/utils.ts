import { Box } from "ivi-shared";
import { OpState } from "./state";
import { Component } from "./component";
import { useSelect } from "./hooks";
import { getDOMNode } from "./reconciler";

/**
 * selector creates a selector factory.
 *
 * @example
 *
 *     interface Entry {
 *       title: string;
 *     }
 *     const useEntryTitle = selector((entry: Entry) => entry.title);
 *
 *     const EntryView = component<Entry>((c) => {
 *       const getEntryTitle = useEntryTitle(c);
 *
 *       return (entry) => div(_, _, getEntryTitle(entry));
 *     });
 *
 * @typename T Selector value type.
 * @param s Selector function.
 * @returns Selector factory.
 */
export function selector<T>(
  s: () => T,
): (component: Component) => () => T;

/**
 * selector creates a selector factory.
 *
 * @example
 *
 * interface Entry {
 *   title: string;
 * }
 *     const useEntryTitle = selector((entry: Entry) => entry.title);
 *
 *     const EntryView = component<Entry>((c) => {
 *       const getEntryTitle = useEntryTitle(c);
 *
 *       return (entry) => div(_, _, getEntryTitle(entry));
 *     });
 *
 * @typename T Selector value type.
 * @param s Selector function.
 * @param shouldUpdate Should update function.
 * @returns Selector factory.
 */
export function selector<T, P>(
  s: (props: P, prev?: T | undefined) => T,
  shouldUpdate?: undefined extends P ? undefined : (prev: P, next: P) => boolean,
): (component: Component) => undefined extends P ? () => T : (props: P) => T;

export function selector<T, P>(
  s: (props: P, prev?: T | undefined) => T,
  shouldUpdate?: undefined extends P ? undefined : (prev: P, next: P) => boolean,
): (component: Component) => undefined extends P ? () => T : (props: P) => T {
  return (component: Component) => useSelect(component, s, shouldUpdate);
}

/**
 * findDOMNode finds a closest DOM node.
 *
 * @typeparam T DOM Node type.
 * @param box Boxed op node state.
 * @returns DOM Node or a `null` value.
 */
export const findDOMNode = <T extends Node>(
  box: Box<OpState | null>,
) => box.v === null ? null : getDOMNode(box.v) as T;
