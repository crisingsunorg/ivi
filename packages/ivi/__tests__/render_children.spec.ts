import { render, checkDOMOps, domOps } from "./utils";
import * as h from "./utils/html";

test(`<div>{ null }</div>`, () => {
  checkDOMOps((c) => {
    const n = render(h.div().c(null));

    expect(n).toMatchSnapshot();
    expect(c).toEqual(domOps(1, 0, 0, 0, 1, 0, 0));
  });
});

test(`<div>"abc"</div>`, () => {
  checkDOMOps((c) => {
    const v = h.div().c("abc");
    const n = render(v);

    expect(n).toMatchSnapshot();
    expect(c).toEqual(domOps(1, 0, 1, 0, 2, 0, 0));
  });
});

test(`<div>10</div>`, () => {
  checkDOMOps((c) => {
    const v = h.div().c(10);
    const n = render(v);

    expect(n).toMatchSnapshot();
    expect(c).toEqual(domOps(1, 0, 1, 0, 2, 0, 0));
  });
});

test(`<div><span></span></div>`, () => {
  checkDOMOps((c) => {
    const v = (
      h.div().c(
        h.span(),
      )
    );
    const n = render(v);

    expect(n).toMatchSnapshot();
    expect(c).toEqual(domOps(2, 0, 0, 0, 2, 0, 0));
  });
});

test(`<div><span></span><strong></strong></div>`, () => {
  checkDOMOps((c) => {
    const v = (
      h.div().c(
        h.span(),
        h.strong(),
      )
    );
    const n = render(v);

    expect(n).toMatchSnapshot();
    expect(c).toEqual(domOps(3, 0, 0, 0, 3, 0, 0));
  });
});

test(`<div><div></div>{ null }<span></span></div>`, () => {
  checkDOMOps((c) => {
    const v = (
      h.div().c(
        h.div(),
        null,
        h.span(),
      )
    );
    const n = render(v);

    expect(n).toMatchSnapshot();
    expect(c).toEqual(domOps(3, 0, 0, 0, 3, 0, 0));
  });
});

test(`<div><div></div>"abc"<span></span></div>`, () => {
  checkDOMOps((c) => {
    const v = (
      h.div().c(
        h.div(),
        "abc",
        h.span(),
      )
    );
    const n = render(v);

    expect(n).toMatchSnapshot();
    expect(c).toEqual(domOps(3, 0, 1, 0, 4, 0, 0));
  });
});

test(`<div><div></div>123<span></span></div>`, () => {
  checkDOMOps((c) => {
    const v = (
      h.div().c(
        h.div(),
        123,
        h.span(),
      )
    );
    const n = render(v);

    expect(n).toMatchSnapshot();
    expect(c).toEqual(domOps(3, 0, 1, 0, 4, 0, 0));
  });
});

test(`complex tree #1`, () => {
  checkDOMOps((c) => {
    const v = (
      h.div().c(
        h.div().c("hello"),
        h.div().c(
          h.span().c("world"),
          h.div().c(
            h.span(),
          ),
        ),
        h.div().c(h.div()),
        h.div(),
      )
    );
    const n = render(v);

    expect(n).toMatchSnapshot();
    expect(c).toEqual(domOps(9, 0, 2, 0, 11, 0, 0));
  });
});
