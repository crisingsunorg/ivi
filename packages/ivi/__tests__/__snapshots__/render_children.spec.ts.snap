// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<div>"abc"</div> 1`] = `
<div>
  abc
</div>
`;

exports[`<div><div></div>"abc"<span></span></div> 1`] = `
<div>
  <div />
  abc
  <span />
</div>
`;

exports[`<div><div></div>{ null }<span></span></div> 1`] = `
<div>
  <div />
  <span />
</div>
`;

exports[`<div><div></div>123<span></span></div> 1`] = `
<div>
  <div />
  123
  <span />
</div>
`;

exports[`<div><span></span></div> 1`] = `
<div>
  <span />
</div>
`;

exports[`<div><span></span><strong></strong></div> 1`] = `
<div>
  <span />
  <strong />
</div>
`;

exports[`<div>{ null }</div> 1`] = `<div />`;

exports[`<div>10</div> 1`] = `
<div>
  10
</div>
`;

exports[`complex tree #1 1`] = `
<div>
  <div>
    hello
  </div>
  <div>
    <span>
      world
    </span>
    <div>
      <span />
    </div>
  </div>
  <div>
    <div />
  </div>
  <div />
</div>
`;
